rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,510 @@
+- The current module system does not provide a way to hook into the usage of GraphQL directives in the operation definition or subgraph schema.
+
+Ultimately, custom modules must be self-contained, composable and testable. They should provide a clear API for developers to interact with the gateway and subgraph lifecycle and implement custom logic without having to understand the internal workings of the router or advanced Go programming concepts.
+To briefly explain the decision to use Go as the language for the module system, we have chosen Go because it is a simple and easy-to-learn language that is widely used in the infrastructure and cloud-native ecosystem. You can build custom integration on top production-grade SDK of AWS, GCP and the community without re-implementing them from scratch. It superiors to scripting languages like Rhai or cross-compiling WebAssembly because it can be easily debugged, profiled, and tested with any modern IDE (VsCode, Goland, etc.). Not part of this RFC, are our ambitions to make the workflow as smooth as possible with a CLI tool that can scaffold, test, and deploy custom modules. In the future, custom modules could be published to a central registry and shared with the community. A brief overview of the workflow is provided at the end of this RFC.
+
+As powerful as the new module becomes, it is important to move basic and common functionality into the core of the router because building and maintaining custom modules should be a last resort. The router should provide a rich set of features out of the box that cover the most common use cases. Custom modules should be reserved for advanced or highly specific use cases that cannot be achieved with the built-in features of the router. Integration with third-party services, custom authentication, and advanced telemetry are examples of use cases that are well-suited for custom modules.
+
+## Proposal
+
+A developer can implement a custom module by creating a struct that implements one or more of the following interfaces:
+
+- `GatewayHooks`: Provides hooks for the gateway lifecycle, including request and response handling.
+- `SubgraphHooks`: Provides hooks for subgraph requests and responses.
+- `ApplicationHooks`: Provides hooks for the application lifecycle, including startup, shutdown, and error handling.
+- `AuthenticationHooks`: Provides hooks for authentication and authorization logic.
+- `TelemetryHooks`: Provides hooks for OpenTelemetry spans and metrics.
+- `OperationHooks`: Provides hooks for parsed, normalized, and planned GraphQL operations.
+- `Module`: Provides hooks for the module lifecycle, including provisioning and shutdown.
+
+```go
+type MyModule struct{}
+
+type Subgraph struct {
+	// The name of the subgraph
+	Name string
+	// The ID of the subgraph
+	ID string
+	// The URL of the subgraph
+	URL *url.URL
+	// The schema of the subgraph
+	Schema *graphql.Schema
+}
+
+type Graph struct {
```

Comment:
Graph is too generic IMO. It's a `FederatedGraph`.

Reply:
We also have monographs. I thought the separation of Graph and Subgraph is clear.

Reply:
And under the hood, a monograph is a federated graph with a single subgraph.

Reply:
Call it Supergraph maybe?

Reply:
We have never called it Supergraph anywhere. Do you still think it is a good idea? I'm also fine with @Aenimus approach to call it FederatedGraph.